client/python/datajunction/base.py

```diff
@@ -0,0 +1,1135 @@
+                    raise DJClientException(  # pragma: no cover
+                        f"Query state {results['state']} is not a DJ-parseable query state!"
+                        " Please reach out to your server admin to make sure DJ is configured"
+                        " correctly.",
+                    )
+
+                # Update the query state and print links if any
+                job_state = models.QueryState(results["state"])
+                if not printed_links and results["links"]:  # pragma: no cover
+                    print(
+                        "Links:\n"
+                        + "\n".join([f"\t* {link}" for link in results["links"]]),
+                    )
+                    printed_links = True
+                progress_bar.title = f"Status: {job_state.value}"
+
+                # Update the polling interval
+                time.sleep(poll_interval)
+                poll_interval *= 2
+
+            # Return results if the job has finished
+            if job_state == models.QueryState.FINISHED:
+                return self.process_results(results)
+            if job_state == models.QueryState.CANCELED:  # pragma: no cover
+                raise DJClientException("Query execution was canceled!")
+            raise DJClientException(  # pragma: no cover
+                f"Error retrieving data: {response.text}",
+            )
+
+    def upsert_materialization(
+        self,
+        node_name: str,
+        config: models.MaterializationConfig,
+    ):
+        """
+        Upserts a materialization config for the node.
+        """
+        response = self._session.post(
+            f"/nodes/{node_name}/materialization/",
+            json=config.dict(),
+        )
+        return response.json()
+
+    def add_availability_state(
+        self,
+        node_name: str,
+        availability: models.AvailabilityState,
+    ):
+        """
+        Adds an availability state for the node
+        """
+        response = self._session.post(
+            f"/data/{node_name}/availability/",
+            json=availability.dict(),
+        )
+        return response.json()
+
+    def set_column_attributes(
+        self,
+        node_name,
+        attributes: List[models.ColumnAttribute],
+    ):
+        """
+        Sets attributes for columns on the node
+        """
+        response = self._session.post(
+            f"/nodes/{node_name}/attributes/",
+            json=[attribute.dict() for attribute in attributes],
+        )
+        return response.json()
+
+
+class ClientEntity(BaseModel):
+    """
+    Any entity that uses the DJ client
+    """
+
+    dj_client: _DJClient = Field(exclude=True)
+
+    class Config:  # pylint: disable=too-few-public-methods
+        """
+        Allow arbitrary types to support _DJClient but exclude
+        it from the output.
+        """
+
+        arbitrary_types_allowed = True
+        exclude = {"dj_client"}
+
+
+class Node(ClientEntity):
```

Comment:
This class and all of the classes after this (Source, Transform, Dimension, etc..) can all start with an underscore, right? The user should never import them and instantiate them directly. They should always go through the client methods.

```py
dj.source(name: str) -> _Source
```

Reply:
Yes, I was considering that and now that you said I'll just do it. I may also rename the `base.py` to `internal.py`. What do you all think?